File: GUILib/src/space/kiibou/net/server/JSONService.java
package space.kiibou.net.server;

import processing.data.JSONObject;
import space.kiibou.net.common.Callbacks;

import java.util.Objects;
import java.util.function.Consumer;

public class JSONService extends Service {

    private final Callbacks<JSONObject, Void> jsonMessageCallbacks;

    public JSONService(Server server) {
        super(server);
        server.registerMessageReceivedCallback(this::onMessageReceived);
        jsonMessageCallbacks = new Callbacks<>();
    }

    @Override
    public void initialize() {

    }

    private void onMessageReceived(long handle, String message) {
        JSONObject parsedMessage = JSONObject.parse(message);

        JSONObject result = new JSONObject();

        result.setLong("handle", handle);
        result.setJSONObject("message", parsedMessage);
    }

    public long registerJSONMessageReceivedCallback(final Consumer<JSONObject> callback) {
        Objects.requireNonNull(callback);
        return jsonMessageCallbacks.addCallback(message -> {
            callback.accept(message);
            return null;
        });
    }

    public void broadcastJSONObject(final JSONObject object) {
        server.broadcastMessage(object.format(-1));
    }

}

File: src/guiLibMain/kotlin/space/kiibou/net/server/service/JsonService.kt
package space.kiibou.net.server.service

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import space.kiibou.net.common.Callbacks
import space.kiibou.net.server.Server
import space.kiibou.net.server.Service

class JsonService(server: Server) : Service(server) {
    private val jsonMessageCallbacks: Callbacks<JsonNode, Unit> = Callbacks()

    val mapper: ObjectMapper = ObjectMapper()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    override fun initialize() {}

    fun registerCallback(callback: (JsonNode) -> Unit): Long {
        return jsonMessageCallbacks.addCallback { message ->
            callback(message)
        }
    }

    fun removeCallback(callbackHandle: Long) {
        jsonMessageCallbacks.removeCallback(callbackHandle)
    }

    fun send(handle: Long, obj: JsonNode) {
        server.sendMessage(handle, mapper.writeValueAsString(obj))
    }

    fun broadcast(obj: JsonNode) {
        server.broadcastMessage(mapper.writeValueAsString(obj))
    }

    init {
        server.registerMessageReceivedCallback { handle, message ->
            val parsedMessage = mapper.readTree(message)
            val result = mapper.createObjectNode()
                .put("handle", handle)
                .set<JsonNode>("message", parsedMessage)
            jsonMessageCallbacks.callAll(result)
        }
    }
}